add PyTorch example (#1)
roscisz/dnn_training_benchmarks: PyTorch_dcgan_lsun/README.md

# PyTorch - DCGAN

This example is based on [this implementation](https://github.com/pytorch/examples/tree/master/dcgan)

## Downloading the dataset

You can download the LSUN dataset by cloning [this repo](https://github.com/fyu/lsun) and running
```

> python download.py -c bedroom

Reviewer: Do we have to download the whole dataset in order to run the example/benchmark?

Author: above command only downloads part of the dataset

Reviewer: lsun (master*) $ du -sh *
32G	bedroom_train_lmdb.zip

We should not require downloading a 32GB dataset only to run a performance benchmark. In the case of our DeepSpeech example, only one training example is used.

BTW. installation of Python request package should be included in the README, maybe through requirements.txt...

Author: The owner doesn't provide an option to download smaller part of the dataset. Should we host a few example images then? I couldn't find any information about the license on their website - https://www.yf.io/p/lsun

Reviewer: Good question... Maybe we should consider another small dataset with clear licensing?

Author: Well, we could use mnist then

Reviewer: The DCGAN implementation used supports some datasets. One of them is "fake":

 dataset = dset.FakeData(image_size=(3, opt.imageSize, opt.imageSize),
                            transform=transforms.ToTensor())

Maybe this would work? We should check it. If it works for the benchmark, we could suggest using it in the README, and also suggest using other datasets to actucally train something.

Author: Added commands for running with fake dataset and updated patch to allow for it to work

### Distributed

There is not that much more to do in this case.

```
from torch import distributed, nn

<argument parser>

parser.add_argument('--backend', type=str, default='gloo', help='Name of the backend to use.')
parser.add_argument(
    '-i',
    '--init-method',
    type=str,
    default='tcp://127.0.0.1:23456',
    help='URL specifying how to initialize the package.')
parser.add_argument('-s', '--world-size', type=int, default=1, help='Number of processes participating in the job.')
parser.add_argument('-r', '--rank', type=int, default=0, help='Rank of the current process.')

if opt.world_size > 1:
        distributed.init_process_group(
            backend=opt.backend,
            init_method=opt.init_method,
            world_size=opt.world_size,
            rank=opt.rank,
        )

<code with model creation, in this case Generator is called netG>
netG = nn.parallel.DistributedDataParallel(netG)

<another model creation in this case discriminator called netD>
netD = nn.parallel.DistributedDataParallel(netD)

> ```

Reviewer: Instead of indications how to change the code, a patch should be provided. See https://github.com/roscisz/dnn_training_benchmarks/blob/master/TensorFlowV1_DeepSpeech_ldc93s1/deepspeech_benchmarking.patch
This way, changes can be introduced easily, and if someone wants to examine and understand them, they will.

Author: But wasn't this a problem with the DeepSpeech example that it tied us to specific version of the used example?

Reviewer: We want our example to be reproducible regardless of changes in the utilized libraries, so we should specify versions for which we checked that it works, and update them as often as possible. This includes version of the training code, so we should provide a specific commit in the DCGAN repo.

The problem with DeepSpeech was that it required a specific version of a binary library hosted by Mozilla, which is no longer hosted. We are temporarily hosting the obsolete version, but updating that example to the latest version seems to be a better solution (BTW I will post an issue regarding that).

Author: Added the patch and changed the instructions accordingly.

               [--beta1 BETA1] [--cuda] [--ngpu NGPU] [--netG NETG]
               [--netD NETD]

distributed arguments:
  --init-method metod,addr, and port to communicate between the nodes [tcp://127.0.0.1:20010]
  --rank current process id
  --world-size number of processes 

optional arguments:
  -h, --help            show this help message and exit
  --dataset DATASET     cifar10 | lsun | mnist |imagenet | folder | lfw | fake
  --dataroot DATAROOT   path to dataset
  --workers WORKERS     number of data loading workers
  --batchSize BATCHSIZE input batch size
  --imageSize IMAGESIZE the height / width of the input image to network
  --nz NZ               size of the latent z vector
  --ngf NGF
  --ndf NDF
  --niter NITER         number of epochs to train for
  --lr LR               learning rate, default=0.0002
  --beta1 BETA1         beta1 for adam. default=0.5
  --cuda                enables cuda
  --ngpu NGPU           number of GPUs to use
  --netG NETG           path to netG (to continue training)
  --netD NETD           path to netD (to continue training)
  --outf OUTF           folder to output images and model checkpoints
  --manualSeed SEED     manual seed

>   --classes CLASSES     comma separated list of classes for the lsun data set

Reviewer: Instead of usage, specific commands would be more useful. See https://github.com/roscisz/dnn_training_benchmarks/tree/master/TensorFlowV1_DeepSpeech_ldc93s1

Author: Maybe leave the usage but add an example command? To suit both type of potential users?

Reviewer: Maybe, but usage is already given in the DCGAN repository and I don't think this redundancy is justified.

# PyTorch - DCGAN

> This example is based on [this implementation](https://github.com/pytorch/examples/tree/master/dcgan)

Reviewer: Maybe providing the proper git clone command would be more handy for the user?

Reviewer: BTW a command for downloading only this example, instead of the whole PyTorch examples repo

Author: As far I was able to find out it isn't possible to clone just one subdirectory of a non-local git repository

Author: done

## Downloading the dataset

> You can download the LSUN dataset by cloning [this repo](https://github.com/fyu/lsun) and running

Reviewer: Same as above

Author: done

# PyTorch - DCGAN

> 

Reviewer: Missing requirements, requirements.txt installation command?

Author: the requirements are in the original repo, and since it needs to be cloned either way I don't see a point in adding them here too

Reviewer: Right, my bad.
Maybe we should mention the requirements installation in README though.

> This example is based on [this implementation](https://github.com/pytorch/examples/tree/master/dcgan)

Reviewer: ```suggestion
This example is based on [the PyTorch DCGAN example](https://github.com/pytorch/examples/tree/master/dcgan).
```

```

> git clone https://github.com/pytorch/examples.git

Reviewer: ```suggestion
git clone https://github.com/pytorch/examples.git
cd examples/
git checkout 0c1654d6913f77f09c0505fb284d977d89c17c1a
```

> cd examples/

Reviewer: ```suggestion
cd examples/dcgan/
```

## Downloading the dataset

> You can download the LSUN dataset by cloning [this repo](https://github.com/fyu/lsun) 

Reviewer: ```suggestion
You can download the LSUN dataset by cloning [the LSUN Demo Code repo](https://github.com/fyu/lsun):
```

```

> git cllone https://github.com/fyu/lsun.git

Reviewer: ```suggestion
git clone https://github.com/fyu/lsun.git
```

```

> and running

Reviewer: ```suggestion
and running the data download script:
```

```

> python download.py -c bedroom

Reviewer: ```suggestion
cd lsun/
python download.py -c bedroom
cd ..
```

```

> For testing and/or benchmarking purposes it is also possible to the train the model on fake dataset. To do this run this command instead of the one given in the Usage section:

Reviewer: ```suggestion
For testing and/or benchmarking purposes it is also possible to the train the model on fake dataset. To do this, set the parameter `--dataset fake` when running the training program.
```

```
CUDA_VISIBLE_DEVICES=0 python main.py --init-method tcp://127.0.0.1:20011 --rank 0 --world 2 --dataset fake --cuda
CUDA_VISIBLE_DEVICES=1 python main.py --init-method tcp://127.0.0.1:20011 --rank 1 --world 2 --dataset fake --cuda

> ```

Reviewer: Remove the commands here and leave them only in the further section?

Author: I think it is easier for the user to copy whole command than to just change a part of it

Reviewer: You're right, but on the other hand:
- we put these modified commands before the regular ones are introduces, which may be confusing
- the user will have to modify the commands to suit their environment anyway
It's your call then.

## Making the code parallel

It is actually very easy to allow PyTorch code to run on multiple GPUs

### Locally

All you need to do to make a data parallel model that can be run on multiple GPUs locally:
```
from torch import nn
<your code including creating the model>
model = nn.DataParallel(model)
```
That's it. After that you just control which GPUs you would like your code to run on by specifing their number in CUDA_VISIBLE_DEVICES before your command (tensorhive takes care of that for you when you specify the devices while creating the task).

### Distributed

> To make the code distributed apply provided patch by running:

Reviewer: ```suggestion
To make the code support distributed training, copy and apply the provided patch:
```

```

> git apply distributed.patch

Reviewer: ```suggestion
cp ../distributed.patch .
git apply distributed.patch
```

```

> You might want to checkout commit 0c1654d6913f77f09c0505fb284d977d89c17c1a to make sure there are no conflicts before applying the patch.

Reviewer: ```suggestion
```

## Usage
```
# set accroding to your configuration
export GLOO_SOCKET_IFNAME=eth0 
```

```

> cd dcgan/

Reviewer: ```suggestion
```

> # set accroding to your configuration

Reviewer: ```suggestion
# set according to your configuration
```

> # Run the following commands simultaneously, on the appropriate hosts:

Reviewer: ```suggestion
# Run the following commands (in the dcgan/ directory) simultaneously, on the appropriate hosts:
```

> CUDA_VISIBLE_DEVICES=0 python main.py --init-method tcp://127.0.0.1:20011 --rank 0 --world 2 --dataset lsun --dataroot <path to lsun dataset> --cuda

Reviewer: Maybe we should indicate that the two commands have to be executed simultaneously on different machines or in different terminals?
If multiple processes are executed on the same machine, we should differentiate CUDA_VISIBLE_DEVICES and tcp ports - maybe we should address this in the README?

Author: CUDA_VISIBLE_DEVICES are different and port should be the same. This example shows running on different devices on the same machine but using techniques that are the same for different machines. I think that for a user looking for distributed benchmarks, this is a common knowledge (to run the commands on different machines if the user want to run the training on them)

Reviewer: You are right about that they know where to run it, but it might looks like the commands should be executed sequentially. I'll add a suggestion for your consideration.

> CUDA_VISIBLE_DEVICES=0 python main.py --init-method tcp://127.0.0.1:20011 --rank 0 --world 2 --dataset lsun --dataroot <path to lsun dataset> --cuda

Reviewer: ```suggestion
cd dcgan/
CUDA_VISIBLE_DEVICES=0 python main.py --init-method tcp://127.0.0.1:20011 --rank 0 --world 2 --dataset lsun --dataroot <path to lsun dataset> --cuda
```

CUDA_VISIBLE_DEVICES=1 python main.py --init-method tcp://127.0.0.1:20011 --rank 1 --world-size 2 --dataset lsun --dataroot <path to lsun dataset> --cuda

> 

Reviewer: ```suggestion
```
used_bboxes += bbox_counts[file]
```

For more details about the usage refer to the [original repo](https://github.com/pytorch/examples/tree/master/dcgan)

> ```

Reviewer: The last sentence should be outside of the ``` environment.

CUDA_VISIBLE_DEVICES=1 python main.py --init-method tcp://127.0.0.1:20011 --rank 1 --world 2 --dataset fake --cuda
```

## Making the code parallel

It is actually very easy to allow PyTorch code to run on multiple GPUs

### Locally

All you need to do to make a data parallel model that can be run on multiple GPUs locally:
```
from torch import nn
<your code including creating the model>
model = nn.DataParallel(model)
```
That's it. After that you just control which GPUs you would like your code to run on by specifing their number in CUDA_VISIBLE_DEVICES before your command (tensorhive takes care of that for you when you specify the devices while creating the task).

### Distributed

To make the code distributed apply provided patch by running:

```
git apply distributed.patch
```

You might want to checkout commit 0c1654d6913f77f09c0505fb284d977d89c17c1a to make sure there are no conflicts before applying the patch.

To learn how to easily adapt PyTorch code for distributed learning please examine the patch. 

## Usage
```
# set accroding to your configuration
export GLOO_SOCKET_IFNAME=eth0 
```

```

> cd dcgan/

Reviewer: ```suggestion
cd dcgan/
# Run the following commands simultaneously, on the appropriate hosts:
```
